src/planning_world.cpp

```diff
@@ -19,229 +20,696 @@ PlanningWorldTpl<S>::PlanningWorldTpl(
+  auto art = articulation_map_.at(name);
+  auto it = planned_articulation_map_.find(name);
+  if (planned && it == planned_articulation_map_.end())
+    planned_articulation_map_[name] = art;
+  else if (!planned && it != planned_articulation_map_.end())
+    planned_articulation_map_.erase(it);
+}
+
+template <typename S>
+std::vector<std::string> PlanningWorldTpl<S>::getNormalObjectNames() const {
+  std::vector<std::string> names;
+  for (const auto &pair : normal_object_map_) names.push_back(pair.first);
+  return names;
+}
+
+template <typename S>
+void PlanningWorldTpl<S>::addPointCloud(const std::string &name,
+                                        const MatrixX3<S> &vertices,
+                                        double resolution) {
+  auto tree = std::make_shared<octomap::OcTree>(resolution);
   for (const auto &row : vertices.rowwise())
-    tree->updateNode(
-        octomap::point3d {static_cast<float>(row(0)), static_cast<float>(row(1)),
-                          static_cast<float>(row(2))},
-        true);
-  point_cloud_ = std::make_shared<CollisionObject>(
-      std::make_shared<fcl::OcTree<S>>(tree), Isometry3<S>::Identity());
-  has_point_cloud_ = true;
+    tree->updateNode(octomap::point3d(row(0), row(1), row(2)), true);
+  auto obj = std::make_shared<CollisionObject>(std::make_shared<fcl::OcTree<S>>(tree));
+  addNormalObject(name, obj);
+}
+
+template <typename S>
+bool PlanningWorldTpl<S>::removeNormalObject(const std::string &name) {
+  auto nh = normal_object_map_.extract(name);
+  if (nh.empty()) return false;
+  attached_body_map_.erase(name);
+  // Update acm_
+  acm_->removeEntry(name);
+  acm_->removeDefaultEntry(name);
+  return true;
+}
+
+template <typename S>
+void PlanningWorldTpl<S>::attachObject(const std::string &name,
+                                       const std::string &art_name, int link_id,
+                                       const Vector7<S> &pose,
+                                       const std::vector<std::string> &touch_links) {
+  auto obj = normal_object_map_.at(name);
+  auto nh = attached_body_map_.extract(name);
+  auto body =
+      std::make_shared<AttachedBody>(name, obj, planned_articulation_map_.at(art_name),
+                                     link_id, toIsometry(pose), touch_links);
+  if (!nh.empty()) {
+    // Update acm_ to disallow collision between name and previous touch_links
+    acm_->removeEntry(name, nh.mapped()->getTouchLinks());
+    nh.mapped() = body;
+    attached_body_map_.insert(std::move(nh));
+  } else
+    attached_body_map_[name] = body;
+  // Update acm_ to allow collision between name and touch_links
+  acm_->setEntry(name, touch_links, true);
+}
+
+template <typename S>
+void PlanningWorldTpl<S>::attachObject(const std::string &name,
+                                       const std::string &art_name, int link_id,
+                                       const Vector7<S> &pose) {
+  auto obj = normal_object_map_.at(name);
```

Comment:
I think maybe there is an even simpler user interface. Suppose the object is already in the scene, we can just ask the user to provide the name of the object and potentially the link the object is attached to. Then, we calculate the relative pose btw the link and the object.